### PR TITLE
[`Attn Masks`] Non-vmap default for attention masks

### LINKED ISSUE #41639: MarianMTModel performance regression due to Bidirectional masks
### System Info

torch 2.10.0.dev20251008+cpu

### Who can help?

@vasqu 

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

`numactl -C 0-3 --membind 0 python test.py`

```python
import time
import torch
from transformers import pipeline, AutoTokenizer

model_id = "Helsinki-NLP/opus-mt-mul-en"
tokenizer = AutoTokenizer.from_pretrained(model_id)
generator = pipeline("translation_fr_to_en", model=model_id, dtype=torch.float16, tokenizer=tokenizer)
generation_config = generator.model.generation_config
generation_config.max_new_tokens = 128
generation_config.min_new_tokens = 128
generation_config.do_sample = False
generation_config.temperature = 1.0
generation_config.num_beams = 1
generation_config.cache_implementation="static"
input_sentence = "C'est fait et soumis. Vous pouvez jouer à « Survival of the Tastiest » sur Android, et sur le web. Le jeu sur le web fonctionne, mais vous devez simuler plusieurs touches pour déplacer les tables, ce qui peut être un peu déroutant. Il y a beaucoup de choses dont j'aimerais parler. Je vais passer en revue chaque sujet, au lieu de faire la liste habituelle de ce qui s'est bien passé et de ce qui s'est mal passé. Concept Travailler sur le thème a probablement été l'une des tâches les plus difficiles auxquelles j'ai dû faire face. A l'origine, j'avais une idée du type de jeu que je voulais développer, du point de vue du gameplay - quelque chose avec beaucoup d'ennemis/acteurs"
for _ in range(5):
    output = generator(input_sentence, generation_config=generation_config)
for _ in range(5):
    start = time.time()
    output = generator(input_sentence, generation_config=generation_config)
    end = time.time()
    print(f"pipeline latency: {(end-start)*1000} ms")
```

### Expected behavior

The performance has 40% regression after PR [41265](https://github.com/huggingface/transformers/pull/41265).

Same issue as #41566 . But the fix PR #41586 only fixes the Bert model regression, and didn't fix MarianMTModel performance regression.

### PR DESCRIPTION
**Pull request by vasqu**
Non-vmap creation of masks. These work with all our base masks and we only default back to vmap when using patterns we cannot guarantee (i.e. additional and/or masks).

Note:
- Non-vmap works with every mask that has anything index based
- Merged old/new sdpa under one function --> easier maintenance imo
- Executorch does not need an additional masking fn anymore
- Lifts some restrictions on older torch versions, e.g. chunked attn with padding, packed attn masks etc

Fixes #41639

cc @jiqing-feng @IlyasMoutawwakil

### CODE REVIEW

File: src/transformers/integrations/executorch.py [hunk 1]
```diff
@@ -1109,92 +1088,3 @@ def _unflatten_dynamic_cache(values, context: torch.utils._pytree.Context):
         value = value_list[idx] if idx < len(value_list) else None
         cache.update(key, value, idx)
     return cache
-
-
-def sdpa_mask_without_vmap(
```
Comment (vasqu) [inline src/transformers/integrations/executorch.py:1114]: No longer needed as vmap was the reason we needed this workaround in the first place

File: src/transformers/masking_utils.py [hunk 2]
```diff
@@ -82,8 +82,10 @@ def causal_mask_function(batch_idx: int, head_idx: int, q_idx: int, kv_idx: int)
 def bidirectional_mask_function(batch_idx: int, head_idx: int, q_idx: int, kv_idx: int) -> bool:
     """
     This creates a full bidirectional mask.
+
+    NOTE: It is important to keep an index-based version for non-vmap expansion.
     """
-    return q_idx.new_ones((), dtype=torch.bool)
+    return q_idx >= 0
```
Comment (vasqu) [inline src/transformers/masking_utils.py:88]: As noted above, for non-vmap we need this as index based version

File: src/transformers/masking_utils.py [hunk 3]
```diff
@@ -402,128 +428,58 @@ def sdpa_mask_recent_torch(
...
+        # We don't need to add an offset to the mask_function either, as we vmap directly the correct indices for k and kv indices
+        with TransformGetItemToIndex():
+            attention_mask = _vmap_for_bhqkv(mask_function)(batch_arange, head_arange, cache_position, kv_arange)
+
+    # Option 3: Limited vmap mask creation (torch<2.6 and custom patterns)
+    else:
+        logger.warning_once(
+            "Using vmap mask creation (custom patterns with and/or masks) has limited capabilities under "
+            "`torch<2.6.0`. We cannot guarantee that the correct mask is constructed."
+        )
```
Comment (vasqu) [inline src/transformers/masking_utils.py]: We shouldn't use vmaps on older torch versions, we still attempt it but warn the users as we cannot guarantee it to work as expected. It all depends on the custom fn they have passed along

File: src/transformers/masking_utils.py [hunk 4]
```diff
@@ -402,128 +428,58 @@ def sdpa_mask_recent_torch(
...
-        causal_mask = causal_mask * padding_mask[:, None, None, :]
+            attention_mask = attention_mask * padding_mask
 
     # Due to a bug in versions of torch<2.5, we need to update the mask in case a query is not attending to any
     # tokens (due to padding). See details in https://github.com/pytorch/pytorch/issues/110213
     if not _is_torch_greater_or_equal_than_2_5 and allow_torch_fix:
-        causal_mask |= torch.all(~causal_mask, dim=-1, keepdim=True)
-    return causal_mask
-
+        attention_mask = attention_mask | torch.all(~attention_mask, dim=-1, keepdim=True)
```
Comment (vasqu) [inline src/transformers/masking_utils.py:435]: I encountered issues with the inplace version where we'd need a clone (e.g. when using swa). This is safer

**Review by ArthurZucker — APPROVED** [review]
can we add a test to default test no graph break on this?

**Comment by jiqing-feng** [comment]
Hi @vasqu . Anything blocks merge?

**Review by Cyrilvallez — APPROVED** [review]
Very very nice, this solves quite a lot of different issues at the same time! I'm very happy to avoid special handling for export! Very clever use of broadcasting from the optimum team, I did not know we could simply do such things! Thanks a lot for upstreaming directly to us.

Do you mind expanding a bit more on what are the limitations of the broadcasting approach here for posterity? Is it only index-based operations as you mention on the comments, or are there more subtle things?

**Comment by IlyasMoutawwakil** [comment]
> Do you mind expanding a bit more on what are the limitations of the broadcasting approach here for posterity? Is it only index-based operations as you mention on the comments, or are there more subtle things?

Nothing I'm aware of, the only condition is to write the mask_function as a comparison between the indexes (and constants).
One example of this is the bidirectional_mask_function which was written as `q_idx.new_ones((), dtype=torch.bool)` simply returns a scalar 1 (true) and @vasqu rewrote it as `q_idx >= 0` (always true). I think any mask function can be written as f(b, h, q, kv) but I can't prove it 😂

**Comment by vasqu** [comment]
Merging this then! Let's see what crazy masks come up in the future; for now the "mask hypothesis" holds :laughing:

**Comment by vasqu** [comment]
@jiqing-feng was mainly blocked by me being out last week ;)
